DOCS-1250: Add sync trigger (#2543)
viamrobotics/docs: docs/data/trigger-sync.md

---

> title: "Trigger Cloud Sync conditionally"

Reviewer: ```suggestion
title: "Trigger Cloud Sync Conditionally"
```

linkTitle: "Trigger Sync"

> description: "Configure cloud sync to automatically capture data in the Viam app."

Reviewer: Should this description mention the conditionality part? Also maybe cloud sync is meant to be capitalized, or else it shouldn't be capitalized in the second sentence of this page.

weight: 36
type: "docs"
tags: ["data management", "cloud", "sync"]
# SME: Alexa Greenberg
---

If you rely on mobile data but have intermittent WiFi connection in certain locations or at certain times of the day, you may want to trigger sync to only occur when these conditions are met.

> This module allows you to configure Cloud Sync to occur only at a specific time frame by implementing a sensor, `naomi:sync-at-time:timesyncsensor`, that enables sync when within a specified timeframe and disables sync when outside that timeframe.

Reviewer: ```suggestion
This module allows you to configure Cloud Sync to occur only at a specific time frame by implementing a sensor, `naomi:sync-at-time:timesyncsensor`, that enables sync within a specified time frame and disables sync outside that time frame.
```
[opt/nits] I think the "when"s are awk and not necessary.
Consider either caps or not for cloud sync, and either time frame or timeframe throughout.

To set up conditional syncing you need to:

1. Create or use an exisitng module that implements a `sensor` component that uses your custom logic to determine when to enable and disable sync.
2. Change the configuration of the data manager to enable selective sync.

## Requirements

> Before configuring your sensor, you must [create a machine](https://docs.viam.com/manage/fleet/machines/#add-a-new-machine) and you also need to:

Author: ```suggestion
Before configuring your sensor, you must [create a machine](https://docs.viam.com/fleet/machines/#add-a-new-machine) and you also need to:
```

1. Enable [data capture](https://docs.viam.com/data/capture/).
2. Enable [cloud sync](https://docs.viam.com/data/cloud-sync/).

## Add sensor to determine when to sync

In this example, you will configure sync to only trigger during a specific time frame of the day using an existing module [`sync-at-time:timesyncsensor`](https://app.viam.com/module/naomi/sync-at-time).
If you need to trigger sync based on a different condition, you need to create your own module and adjust the module logic accordingly.
Additional examples are available in this [GitHub Repo](https://github.com/viam-labs/trigger-sync-examples-v2).

To use [`sync-at-time:timesyncsensor`](https://app.viam.com/module/naomi/sync-at-time):

> 1. Go to your machine's **Config** page and click **Add Component**.

Reviewer: ```suggestion
1. Go to your machine's **Config** page and click **Create component**.
```

2. Then select the `sync-at-time:timesyncsensor` model from the [`sync-at-time` module](https://app.viam.com/module/naomi/sync-at-time).
3. Click **Add module**, then enter a name for your sensor and click **Create**.
   The sensor will return true and enable sync when in a specified time frame.

> 4. To configure your timeframe, go to the new component panel and copy and paste the following attribute template into your sensor’s **Attributes** box:

Reviewer: ```suggestion
4. To configure your time frame, go to the new component panel and copy and paste the following attribute template into your sensor’s **Attributes** box:
```
[nit] For consistency. Spelled with two words in most of this doc, which Google tells me is the most common nowadays

>    {{< tabs >}}
>    {{% tab name="Config builder" %}}

Reviewer: ```suggestion
{{< tabs >}}
{{% tab name="Config builder" %}}
```
Maybe? Or indent the other ones to match?? Unsure

Author: indent is to ensure it's indented for the bullet it is under

Reviewer: Yeah I just thought the closing tags of the tabs should be aligned with the opening ones

```json
{
  "start": "HH:MM:SS",
  "end": "HH:MM:SS",
  "zone": "<TIMEZONE>"
}
```

{{% /tab %}}
{{% tab name="JSON Template" %}}

```json
{
  "name": "<SENSORNAME>",
  "model": "naomi:sync-at-time:timesyncsensor",
  "type": "sensor",
  "namespace": "rdk",
  "attributes": {
    "start": "HH:MM:SS",
    "end": "HH:MM:SS",
    "zone": "<TIMEZONE>"
  },
  "depends_on": []
}
```

{{% /tab %}}
{{% tab name="JSON Example" %}}

```json
{
  "name": "timesensor",
  "model": "naomi:sync-at-time:timesyncsensor",
  "type": "sensor",
  "namespace": "rdk",
  "attributes": {
    "start": "18:29:00",
    "end": "18:30:00",
    "zone": "CET"
  },
  "depends_on": []
}
```

{{% /tab %}}
{{< /tabs >}}

The following attributes are available for the `naomi:sync-at-time:timesyncsensor` sensor:

<!-- prettier-ignore -->
| Name    | Type   | Inclusion    | Description |
| ------- | ------ | ------------ | ----------- |

> | `start` | string | **Required** | The start time for the time frame during which you want to sync, example: `"14:10:00"`.  |

Reviewer: ```suggestion
| `start` | string | **Required** | The start time for the time frame during which you want to sync, for example: `"14:10:00"`.  |
```
OR
```suggestion
| `start` | string | **Required** | The start time for the time frame during which you want to sync. Example: `"14:10:00"`.  |
```
[opt]

> | `end`   | string | **Required** | The start time for the time frame during which you want to sync, example: `"15:35:00"`. |

Reviewer: ```suggestion
| `end`   | string | **Required** | The end time for the time frame during which you want to sync, for example: `"15:35:00"`. |
```
the "for" is opt but I think without it there should be a period instead of a comma

Reviewer: ```suggestion
| `end`   | string | **Required** | The end of the sync time frame, for example: `"15:35:00"`. |
```
[opt] Could also shorten both of these

Reviewer: Also fix "start" in the readme

| `zone`  | string | **Required** | The time zone for the `start` and `end` time, for example: `"CET"`. |

In the next step you will configure the data manager to take the sensor into account when syncing.

### Configure data manager to sync based on sensor

On your machine's **Config** tab, switch to **JSON** mode and add a `selective_syncer_name` with the name for the sensor you configured:

{{< tabs >}}
{{% tab name="JSON Template" %}}

> ```json {class="line-numbers linkable-line-numbers" data-line="6"}

Reviewer: Love the use of line highlighting

{

>   "attributes": {
>     "additional_sync_paths": [],
>     "capture_dir": "",
>     "capture_disabled": false,
>     "selective_syncer_name": "<SENSOR-NAME>",
>     "sync_disabled": false,
>     "sync_interval_mins": 0.1,
>     "tags": []
>   },
>   "name": "Data-Management-Service",
>   "type": "data_manager"

Reviewer: ```suggestion
  "name": "Data-Management-Service",
  "type": "data_manager",
  "attributes": {
    "additional_sync_paths": [],
    "capture_dir": "",
    "capture_disabled": false,
    "selective_syncer_name": "<SENSOR-NAME>",
    "sync_disabled": false,
    "sync_interval_mins": 0.1,
    "tags": []
  }
```
[opt] This is the order in which it shows up on https://docs.viam.com/data/capture/#add-the-data-management-service so people can recognize it easily,  and is more intuitive

}
```

{{% /tab %}}
{{% tab name="JSON Example" %}}

```json {class="line-numbers linkable-line-numbers" data-line="6"}
{

>   "attributes": {
>     "additional_sync_paths": [],
>     "capture_dir": "",
>     "capture_disabled": false,
>     "selective_syncer_name": "timesensor",
>     "sync_disabled": false,
>     "sync_interval_mins": 0.1,
>     "tags": []
>   },
>   "name": "Data-Management-Service",
>   "type": "data_manager"

Reviewer: ```suggestion
  "name": "Data-Management-Service",
  "type": "data_manager",
  "attributes": {
    "additional_sync_paths": [],
    "capture_dir": "",
    "capture_disabled": false,
    "selective_syncer_name": "timesensor",
    "sync_disabled": false,
    "sync_interval_mins": 0.1,
    "tags": []
  }
```
as above

}
```

{{% /tab %}}
{{< /tabs >}}

{{% expand "Click to view a full configuration example" %}}

> ```json

Reviewer: [opt] Consider adding line highlighting to the whole cloud sync section

Author: ```suggestion
```json {class="line-numbers linkable-line-numbers" data-line="7,12"}
```

      "type": "sensor",
      "namespace": "rdk"
    }
  ],
  "services": [
    {
      "attributes": {
        "selective_syncer_name": "timesensor",
        "sync_interval_mins": 0.2,
        "capture_dir": "",
        "tags": [],
        "additional_sync_paths": []
      },
      "name": "datamanager",
      "type": "data_manager",
      "namespace": "rdk"
    }
  ],
  "agent_config": {
    "subsystems": {
      "viam-server": {
        "release_channel": "stable",
        "pin_version": "",
        "pin_url": "",
        "disable_subsystem": false
      },
      "agent-provisioning": {
        "release_channel": "stable",
        "pin_version": "",
        "pin_url": "",
        "disable_subsystem": false
      },
      "viam-agent": {
        "release_channel": "stable",
        "pin_version": "",
        "pin_url": "",
        "disable_subsystem": false
      }
    }
  },
  "modules": [
    {
      "module_id": "naomi:sync-at-time",
      "version": "0.0.2",
      "type": "registry",
      "name": "naomi_sync-at-time"
    }
  ]
}
```

{{% /expand%}}

You have now configured sync to happen during a specific time slot.

## Test your sync configuration

> To test your setup, [configure a webcam](https://docs.viam.com/components/camera/webcam/) or another component and [enable data capture on one of the component methods](https://docs.viam.com/data/capture/#configure-data-capture-for-individual-components), for example `ReadImage`.

Reviewer: ```suggestion
To test your setup, [configure a webcam](https://docs.viam.com/components/camera/webcam/) or another component and [enable data capture on the component](https://docs.viam.com/data/capture/#configure-data-capture-for-individual-components).
```
[opt] I think of it as data capture being configured on the component, with the methods as config options but not as the objects upon which data capture is configured.

Author: hmm, I do want to specifically mention `readimage` because get images doesn't work/causes an error

Author: ```suggestion
To test your setup, [configure a webcam](https://docs.viam.com/components/camera/webcam/) or another component and [enable data capture on the component](https://docs.viam.com/data/capture/#configure-data-capture-for-individual-components).
For a camera component, use the `ReadImage` method.
```

The data manager will now capture data.
Go to the [**Control** tab](https://docs.viam.com/fleet/machines/#control).
You should see the sensor.
Click on `GetReadings`.

{{<imgproc src="/data/timesensor.png" resize="800x" declaredimensions=true alt="Control tab with sensor panel">}}

> If you are in the time frame for sync, the sensor will return true.

Reviewer: ```suggestion
If you are in the time frame for sync, the time sync sensor will return true.
```
[opt] Maybe overkill but in case someone is syncing sensor data, to disambiguate

You can confirm that no data is currently syncing by going to the [**Data** tab](https://app.viam.com/data/view).

> If you are not in the time frame for sync, adjust the configuration of your sensor.

Reviewer: [opt]
```suggestion
If you are not in the time frame for sync, adjust the configuration of your time sync sensor.
```

---
title: "Trigger Cloud Sync Conditionally"
linkTitle: "Trigger Sync"

> description: "Configure cloud sync to automatically capture data in the Viam app."

Author: ```suggestion
description: "Trigger cloud sync to sync captured data conditionally."
```
